x-pack/solutions/security/plugins/security_solution/public/data_view_manager/redux/listeners/data_view_selected.ts

```diff
@@ -0,0 +1,108 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { DataView, DataViewsServicePublic } from '@kbn/data-views-plugin/public';
+import type { AnyAction, Dispatch, ListenerEffectAPI } from '@reduxjs/toolkit';
+import isEmpty from 'lodash/isEmpty';
+import type { RootState } from '../reducer';
+import { scopes } from '../reducer';
+import { selectDataViewAsync } from '../actions';
+import { sharedDataViewManagerSlice } from '../slices';
+
+export const createDataViewSelectedListener = (dependencies: {
+  dataViews: DataViewsServicePublic;
+}) => {
+  return {
+    actionCreator: selectDataViewAsync,
+    effect: async (
+      action: ReturnType<typeof selectDataViewAsync>,
+      listenerApi: ListenerEffectAPI<RootState, Dispatch<AnyAction>>
+    ) => {
+      let dataViewByIdError: unknown;
+      let adhocDataViewCreationError: unknown;
+      let dataViewById: DataView | null = null;
+      let adHocDataView: DataView | null = null;
+
+      const state = listenerApi.getState();
+
+      const findCachedDataView = (id: string | null | undefined) => {
+        if (!id) {
+          return null;
+        }
+
+        const cachedAdHocDataView =
+          state.dataViewManager.shared.adhocDataViews.find((dv) => dv.id === id) ?? null;
+
+        const cachedPersistedDataView =
+          state.dataViewManager.shared.dataViews.find((dv) => dv.id === id) ?? null;
+
+        const cachedDataView = cachedAdHocDataView || cachedPersistedDataView;
+
+        // NOTE: validate if fields are available, otherwise dont return the view
+        // This is required to compute browserFields later.
+        // If the view is not returned here, it will be fetched further down this file, and that
+        // should return the full data view.
+        if (isEmpty(cachedDataView?.fields)) {
+          return null;
+        }
+
+        return cachedDataView;
+      };
+
+      /**
+       * Try to locate the data view in cached entries first
+       */
+      const cachedDataViewSpec = findCachedDataView(action.payload.id);
+
+      if (!cachedDataViewSpec) {
+        try {
+          if (action.payload.id) {
+            dataViewById = await dependencies.dataViews.get(action.payload.id);
+          }
+        } catch (error: unknown) {
+          dataViewByIdError = error;
+        }
+      }
+
+      if (!dataViewById) {
+        try {
+          const title = action.payload.fallbackPatterns?.join(',') ?? '';
+          if (!title.length) {
+            throw new Error('empty adhoc title field');
+          }
+
+          adHocDataView = await dependencies.dataViews.create({
+            id: `adhoc_${title}`,
+            title,
+          });
+          if (adHocDataView) {
+            listenerApi.dispatch(sharedDataViewManagerSlice.actions.addDataView(adHocDataView));
```

Comment:
what's the best way to test adhoc dataview?

Reply:
I would use the old timelines from [this PR ](https://github.com/elastic/kibana/pull/120022)(the link to the JSON there)...those have 7.x timelines with index patterns that don't exist and should create an adhoc dataview when opened in timeline. Otherwise you can create them via the create dataview ui in discover

Reply:
thanks @michaelolo24! yes, create a data view and choose `use without saving` also creates an ad hoc data view

![image](https://github.com/user-attachments/assets/4d850eac-04d1-4310-8165-9b44880f3519)